File: server/src/com/thoughtworks/go/server/controller/AuthorizationController.java
/*************************GO-LICENSE-START*********************************
 * Copyright 2014 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *************************GO-LICENSE-END***********************************/

package com.thoughtworks.go.server.controller;

import com.thoughtworks.go.i18n.Localizer;
import com.thoughtworks.go.plugin.access.authentication.AuthenticationPluginRegistry;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.servlet.ModelAndView;

import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.HashMap;

@Controller
public class AuthorizationController {
    private final Localizer localizer;
    private AuthenticationPluginRegistry authenticationPluginRegistry;

    @Autowired
    public AuthorizationController(Localizer localizer, AuthenticationPluginRegistry authenticationPluginRegistry) {
        this.localizer = localizer;
        this.authenticationPluginRegistry = authenticationPluginRegistry;
    }

    @RequestMapping(value = "/auth/login", method = RequestMethod.GET)
    public ModelAndView login(@RequestParam(value = "login_error", required = false) Boolean loginError,
                              HttpServletResponse response) throws IOException {

        response.setHeader("Cache-Control", "no-cache, must-revalidate, no-store");
        response.setHeader("Pragma", "no-cache");

        HashMap model = new HashMap();
        model.put("login_error", loginError);
        model.put("l", localizer);
        model.put("authentication_plugin_registry", authenticationPluginRegistry);
        return new ModelAndView("auth/login", model);
    }

    @RequestMapping(value = "/auth/security_check", method = RequestMethod.POST)
    public ModelAndView securityCheckHandlerWhenAuthenticationProcessingFilterIsOff(HttpServletResponse response)
            throws IOException {
        response.sendRedirect("/go");
        return null;
    }
}

File: server/webapp/WEB-INF/rails/app/controllers/api_v1/admin/templates/authorization_controller.rb
#
# Copyright 2019 ThoughtWorks, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

module ApiV1
  module Admin
    module Templates
      class AuthorizationController < BaseController

        before_action :check_admin_user_and_403
        before_action :load_template, only: [:show, :update]
        before_action :check_for_stale_request, only: [:update]

        def show
          json = ApiV1::Admin::Authorization::AuthorizationConfigRepresenter.new(@template.getAuthorization).to_hash(url_builder: self)
          render DEFAULT_FORMAT => json if stale?(strong_etag: etag_for(@template))
        end

        def update
          result = HttpLocalizedOperationResult.new
          updated_authorization = ApiV1::Admin::Authorization::AuthorizationConfigRepresenter.new(com.thoughtworks.go.config.Authorization.new).from_hash(params[:authorization])
          template_config_service.updateTemplateAuthConfig(current_user, @template, updated_authorization, result, etag_for(@template))
          handle_update_response(result, updated_authorization)
        end

        private
        def load_template(template_name = params[:template_name])
          result = HttpLocalizedOperationResult.new
          @template = template_config_service.loadForView(template_name, result)
          raise RecordNotFound unless @template
        end

        def handle_update_response(result, updated_authorization)
          json = ApiV1::Admin::Authorization::AuthorizationConfigRepresenter.new(updated_authorization).to_hash(url_builder: self)
          if result.isSuccessful
            response.etag = [etag_for(@template)]
            render DEFAULT_FORMAT => json
          else
            render_http_operation_result(result, {data: json})
          end
        end

        def etag_for_entity_in_config
          etag_for(@template)
        end

        def stale_message
          com.thoughtworks.go.i18n.LocalizedMessage::staleResourceConfig('Template', @template.name)
        end
      end
    end
  end
end
